### PR TITLE
Bump go version to 1.8.3

### PR DESCRIPTION
**Pull request by cyli**
The patches no longer merged cleanly - luckily @tiborvass figured out how to get the first patch to apply.  :)

### CODE REVIEW

File: patches/0001-Expose-Encoder.Canonical-and-MarshalCanonical-Handle.patch [hunk 1]
```diff
@@ -380,21 +377,15 @@ index fca2a09..aaa79c2 100644
 +	} else {
 +		x.byIndex = f
 +	}
++	newM[t] = x
+ 	fieldCache.value.Store(newM)
```
Comment (cyli) [inline patches/0001-Expose-Encoder.Canonical-and-MarshalCanonical-Handle.patch:415]: @tiborvass found that The previous code did not actually do anything with `x` - it did not set it back to `fieldCache`, which was a bug.  When encoding JSON, it seems like some fields didn't actually get sorted by name.

So we ended up with serialized values like:  `{"keytype":"ecdsa-x509","keyval":{"private":null,"public":"..."}}` where the keys in `"keyval"` were not lexicographically ordered.

The updated code does set `x` back into the `fieldCache`, so we end up with items like: `{"keytype":"ecdsa-x509","keyval":{"public":"...","private":null}}`, which is correctly lexicographically ordered.

This may cause backwards compatibility issues in users of previous versions of canonical JSON.   cc @endophage @riyazdf @ecordell

Comment (endophage) [reply patches/0001-Expose-Encoder.Canonical-and-MarshalCanonical-Handle.patch:415]: Am I missing something? `... q, r, s, t, u ...`. Doesn't `private` come before `public` so `{"keytype":"ecdsa-x509","keyval":{"public":"...","private":null}}`, is _not_ correctly lexicographically ordered.

Comment (cyli) [reply patches/0001-Expose-Encoder.Canonical-and-MarshalCanonical-Handle.patch:415]: @endophage Blah, you are correct.  At 1am for some reason I kept reading the "u" as "o".  I'll see if I can add to a test that duplicates this particular issue.

Comment (tiborvass) [reply patches/0001-Expose-Encoder.Canonical-and-MarshalCanonical-Handle.patch:415]: Yeah let's add a test that stresses this codepath.

**Comment by cyli** [comment]
Update:  I have no idea what's causing the notary issues, since adding the specific key marshaling struct from notary in the test succeeded, so I thought it had to do with caching.  So I updated the marshaling test to first marshal the `CanonicalTestStruct` in a non-canonical manner, and then marshal them again canonically, and a bunch of the tests fail.  So trying to track that down ATM.  Partially it has to do with `encoderCache` storing the non-canonical marshaling functions.

**Comment by cyli** [comment]
@tiborvass With [this change](https://github.com/docker/go/pull/6/commits/b8d65865e24f675e7546a1ac5d00ed0f9750c9fd), and a small fix to notary, all tests pass.

Could you see if it makes sense?

**Comment by thaJeztah** [comment]
ping @tiborvass @endophage PTAL - we also need to update this for Go 1.8.5 (and now for Go 1.9.3)

**Comment by justincormack** [comment]
Closing, we should obsolete this, and tracking upstream is not that useful.
